### PR TITLE
Fix cross-model dashboards

### PR DESCRIPTION
**Pull request by rbarry82**

### CODE REVIEW

**Review by balbirthomas — APPROVED** [review]
The code changes look clean too me. However I am only minimally familiar with context here so can not evaluate the deeper ramifications of this change. Perhaps a commit message providing context may be useful.

**Review by mmanciop — DISMISSED** [review]

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 1]
```diff
@@ -89,7 +118,8 @@ def __init__(
         self,
         charm: CharmBase,
         name: str,
-        event_relation: str = "monitoring",
+        event_relation: Optional[str] = "monitoring",
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:121]: Based on the code in this PR's diff it seems that `None` is not a valid value. Drop the Optional?
```suggestion
        event_relation = "monitoring",
```

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 2]
```diff
@@ -115,15 +145,22 @@ def __init__(
...
                 restore the dashboard
+            dashboards_path: a filesystem path relative to the charm root
+                where dashboard templates can be located
         """
         super().__init__(charm, name)
         self.charm = charm
         self.name = name
+        self._DASHBOARDS_PATH = dashboards_path
         self._stored.set_default(dashboards={}, dashboard_templates={}, event_relation=None)
         self._stored.event_relation = event_relation
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:156]: Duplicated assignment
```suggestion
        self._stored.set_default(dashboards={}, dashboard_templates={}, event_relation=event_relation)
```

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 3]
```diff
@@ -115,15 +145,22 @@ def __init__(
...
 
         events = self.charm.on[name]
 
+        self.framework.observe(self.charm.on.upgrade_charm, self._on_upgrade_charm)
+
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:162]: Consider mentioning in the docstring that this is happening behind the scenes.

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 4]
```diff
@@ -137,30 +174,63 @@ def __init__(
...
+        """Watch for a relation being joined and automatically send dashboards.
+
+        Args:
+            rel: The :class:`Relation` to set grafana_dashboard data for
+        """
+        data = {}
+
+        for path in Path(self._DASHBOARDS_PATH).glob("*.tmpl"):
+            if not path.is_file():
+                continue
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:206]:
```suggestion
        for path in filter(Path.is_file, Path(self._DASHBOARDS_PATH).glob("*.tmpl")):
```

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 5]
```diff
@@ -263,31 +327,38 @@ def invalidate_dashboard(self, reason: str, rel_id=None) -> None:
...
-            event.defer()
-            return
-
-        dash = self._stored.dashboard_templates.get(rel.id, None)
-        if dash:
-            dash_tmpl = zlib.decompress(base64.b64decode(dash.encode())).decode()
+        for relation in self.charm.model.relations[self._stored.event_relation]:
+            if len(relation.units) == 0:
+                event.defer()
+                return
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:346]: This is a very interesting construct.
Can you add a comment explaining the rationale?
Are you waiting for _all_ relations to have at least one unit? Why?

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 6]
```diff
@@ -263,31 +327,38 @@ def invalidate_dashboard(self, reason: str, rel_id=None) -> None:
...
+                self._update_dashboards(type_convert_stored(dash), dash_relation.id)
 
     def _on_monitoring_relation_broken(self, _) -> None:
         """Invalidate the dashboard if there are no more relations."""
-        if len(self.model.get_relation(self._stored.event_relation).units) == 0:
+        valid = False
+        for relation in self.charm.model.relations[self._stored.event_relation]:
+            # Force an actual lookup to update `RelationMapping`
+            if len(relation.units) > 0:
+                valid = True
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:359]: add `break`?

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 7]
```diff
@@ -398,32 +493,40 @@ def _validate_dashboard_data(self, data: Dict, rel: Relation) -> None:
         if not grafana_datasource:
             return
 
+        # Import at runtime so we don't get client dependencies
```
Comment (sed-i) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:496]: I don't understand this. What do you mean by "client dependency"? Why here is "at runtime" compared to top of file?

File: src/charm.py [hunk 8]
```diff
@@ -144,6 +146,15 @@ def _on_grafana_source_changed(self, event: GrafanaSourceEvents) -> None:
...
         self._configure(event)
 
+    def _on_upgrade_charm(self, event: UpgradeCharmEvent) -> None:
+        """Re-provision Grafana and its datasources on upgrade.
+
+        Args:
+            event: a :class:`UpgradeCharmEvent` to signal the upgrade
+        """
+        self._configure(event)
+        self._on_dashboards_changed(event)
```
Comment (sed-i) [inline src/charm.py:156]: Would it make sense to move this into the provider lib, similar to what you have with the consumer?
Unless `_configure` must happen before..?

**Review by sed-i — APPROVED** [review]

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 9]
```diff
@@ -25,11 +38,27 @@
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 1
+LIBPATCH = 4
```
[the remark this answers is not in the document]
Comment (mmanciop) [reply lib/charms/grafana_k8s/v0/grafana_dashboard.py:41]: There were a few versions in between that we tested and we need to make sure ;-)

**Comment by mmanciop** [comment]
@sed-i I did not see your comments before merging this, I'll address them in a PR I am working on to streamline the constructors

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py (hunk 5 above)
Comment (mmanciop) [reply lib/charms/grafana_k8s/v0/grafana_dashboard.py:346]: It is waiting for a Prometheus unit to be there to actually provide data to power the dashboard.

**Pull request merged**